source/extensions/common/wasm/wasm.cc

```diff
@@ -28,6 +33,55 @@ namespace Common {
 namespace Wasm {
 namespace {
 
+Secret::GenericSecretConfigProviderSharedPtr
+secretsProvider(const envoy::extensions::transport_sockets::tls::v3::SdsSecretConfig& config,
+                Secret::SecretManager& secret_manager,
+                Server::Configuration::TransportSocketFactoryContext& transport_socket_factory,
+                Init::Manager& init_manager) {
+  if (config.has_sds_config()) {
+    return secret_manager.findOrCreateGenericSecretProvider(config.sds_config(), config.name(),
+                                                            transport_socket_factory, init_manager);
+  } else {
+    return secret_manager.findStaticGenericSecretProvider(config.name());
+  }
+}
+
+absl::StatusOr<std::string> getRegistryCredential(
+    envoy::extensions::wasm::v3::VmConfig& vm_config, Upstream::ClusterManager& cluster_manager,
+    Server::Configuration::TransportSocketFactoryContext& transport_socket_factory,
+    Init::Manager& init_manager, ThreadLocal::SlotAllocator& slot_alloc, Api::Api& api,
+    std::string& registry) {
+  if (!vm_config.has_image_pull_secret()) {
+    return "";
+  }
+
+  auto image_pull_secret_provider = secretsProvider(
+      vm_config.image_pull_secret(), cluster_manager.clusterManagerFactory().secretManager(),
+      transport_socket_factory, init_manager);
+  if (image_pull_secret_provider == nullptr) {
+    return absl::FailedPreconditionError("Secret provider for Wasm plugin is null");
+  }
+
+  auto secret_reader_result = Secret::ThreadLocalGenericSecretProvider::create(
```

Comment:
A better pattern is we store the thread local provider as a shared ptr, and then call `secret_reader->secret()` during `createWasm()` rather than creating it each time. The thread local secret can be updated by SDS, and it can then unblock some retry usage with newly updated secret.

Reply:
It's good that you raised this issue, because I was wondering if caching the WASM code makes sense taking into account that we specify the expected SHA. This means when the cache expires and Envoy fetches the WASM code again, it still expects the same SHA specified in the config, so what's the point in refreshing the plugin?

Reply:
Done.

Reply:
Please let me know if the implementation is good enough to implement unit tests. So far, I was only testing this feature manually with a real registry and a wasm plugin, as I wanted to avoid rewriting unit tests back and forth.